src/process/process_set.cu

```diff
@@ -3,288 +3,223 @@
       cudaMemcpy(d_jacobian, sparseMatrix.jacobian_, sizeof(double) * sparseMatrix.jacobian_size_, cudaMemcpyHostToDevice);
-      cudaMemcpy(
-          d_number_of_reactants,
-          processSet.number_of_reactants_,
-          sizeof(size_t) * matrixParam.n_reactions_,
-          cudaMemcpyHostToDevice);
-      cudaMemcpy(
-          d_reactant_ids, processSet.reactant_ids_, sizeof(size_t) * processSet.reactant_ids_size_, cudaMemcpyHostToDevice);
-      cudaMemcpy(
-          d_number_of_products,
-          processSet.number_of_products_,
-          sizeof(size_t) * matrixParam.n_reactions_,
-          cudaMemcpyHostToDevice);
-      cudaMemcpy(d_yields, processSet.yields_, sizeof(double) * processSet.yields_size_, cudaMemcpyHostToDevice);
-      cudaMemcpy(
-          d_jacobian_flat_ids,
-          processSet.jacobian_flat_ids_,
-          sizeof(size_t) * processSet.jacobian_flat_ids_size_,
-          cudaMemcpyHostToDevice);
-      cudaMemcpy(&(device->rate_constants_), &d_rate_constants, sizeof(double*), cudaMemcpyHostToDevice);
-      cudaMemcpy(&(device->state_variables_), &d_state_variables, sizeof(double*), cudaMemcpyHostToDevice);
-      cudaMemcpy(&(device->jacobian_), &d_jacobian, sizeof(double*), cudaMemcpyHostToDevice);
-      cudaMemcpy(&(device->number_of_reactants_), &d_number_of_reactants, sizeof(size_t*), cudaMemcpyHostToDevice);
-      cudaMemcpy(&(device->reactant_ids_), &d_reactant_ids, sizeof(size_t*), cudaMemcpyHostToDevice);
-      cudaMemcpy(&(device->number_of_products_), &d_number_of_products, sizeof(size_t*), cudaMemcpyHostToDevice);
-      cudaMemcpy(&(device->yields_), &d_yields, sizeof(double*), cudaMemcpyHostToDevice);
-      cudaMemcpy(&(device->jacobian_flat_ids_), &d_jacobian_flat_ids, sizeof(size_t*), cudaMemcpyHostToDevice);
 
       // setup kernel
-      size_t total_blocks = (matrixParam.n_grids_ + BLOCK_SIZE - 1) / BLOCK_SIZE;
-
+      size_t num_blocks = (matrixParam.n_grids_ + BLOCK_SIZE - 1) / BLOCK_SIZE;
       size_t n_reactions = matrixParam.n_reactions_;
       size_t n_grids = matrixParam.n_grids_;
+
       // launch kernel and measure time performance
       auto startTime = std::chrono::high_resolution_clock::now();
-      AddJacobianTermsKernel<<<total_blocks, BLOCK_SIZE>>>(device, n_grids, n_reactions);
+      FormJacobianMatrixKernel<<<num_blocks, BLOCK_SIZE>>>(d_rate_constants, d_state_variables, d_jacobian,
+                                                           devstruct, n_grids, n_reactions);
       cudaDeviceSynchronize();
       auto endTime = std::chrono::high_resolution_clock::now();
       auto kernel_duration = std::chrono::duration_cast<std::chrono::nanoseconds>(endTime - startTime);
 
+      // copy the data from device to host
       cudaMemcpy(sparseMatrix.jacobian_, d_jacobian, sizeof(double) * sparseMatrix.jacobian_size_, cudaMemcpyDeviceToHost);
+
       // clean up
       cudaFree(d_rate_constants);
       cudaFree(d_state_variables);
       cudaFree(d_jacobian);
```

Comment:
Do these need to be allocated and freed each time? Couldn't we also allocate space for this at the beginning and just overwrite the data at the beginning of this kernel?

Reply:
Agreed. It is more efficient to do what you suggested. However, there variables are either input variables for MICM or temporary matrix-type variables. Thus I think it should be handled in the `CudaMatrix` class or at the beginning of the MICM solver instead.